### PR TITLE
Fix EC2 discovery settings

### LINKED ISSUE #18652: AWS IAM role not working with Elasticsearch 5.0.0-alpha2, but works with 2.3 version
**Elasticsearch version**:  5.0.0-alpha2

**JVM version**: 1.8.0_91(1.8.0_91-b14)

**OS version**: Red Hat Enterprise Linux 7.2 (3.10.0-327.18.2.el7.x86_64)

**Description of the problem including expected versus actual behavior**:
I am trying to use AWS IAM role with Elasticsearch 5.0.0-alpha2 and EC2 discovery plug-in, but it does not seem to be working and I am getting below error:

> "Exception while retrieving instance list from AWS API: Authorization header or parameters are not formatted correctly. (Service: AmazonEC2; Status Code: 401; Error Code: AuthFailure"

I am using below configuration with jdk8:

> cluster.name: "test-cluster"
> cloud.aws.region: "us-west-2"
> cloud.aws.ec2.region: "us-west-2"
> cloud.aws.ec2.protocol: "http"
> discovery.type: "ec2"
> #bootstrap.mlockall: true
> node.master: true
> node.data: false
> node.name: ${HOSTNAME}-Master
> discovery.zen.minimum_master_nodes: 1
> network.host: ec2:privateIp
> discovery.ec2.any_group: true
> discovery.ec2.groups : sg-9d856tfe

And, below is IAM role permission that I have configured with elasticsearch instance:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "ec2:Describe*"
            ],
            "Effect": "Allow",
            "Resource": [
                "*"
            ]
        }
    ]
}
```

However, same configuration works fine with with Elasticsearch 2.3 version. Has anyone faced the same issue with the latest Elasticsearch version 5.0.0-alpha2?

Also, today I enabled the debug mode of AWS call and I could notice(see below) that its loading credentials from **StaticCredentialsProvider** - its wrong behavior. As access and secret key is absent in the config(elasticsearch.yml) file, so ideally it should load credential from **InstanceProfileCredentialsProvider**.

> [DEBUG][com.amazonaws.auth.AWSCredentialsProviderChain] Loading credentials from com.amazonaws.internal.StaticCredentialsProvider@40bf7b26

**Steps to reproduce**:
 1. Start elasticsearch master data node with the above mentioned configuration.

**Provide logs (if relevant)**:
```
com.amazonaws.AmazonServiceException: Authorization header or parameters are not formatted correctly. (Service: AmazonEC2; Status Code: 401; Error Code: AuthFailure; Request ID: )
        at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1239)
        at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:823)
        at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:506)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:318)
        at com.amazonaws.services.ec2.AmazonEC2Client.invoke(AmazonEC2Client.java:11901)
        at com.amazonaws.services.ec2.AmazonEC2Client.describeInstances(AmazonEC2Client.java:5940)
        at org.elasticsearch.discovery.ec2.AwsEc2UnicastHostsProvider.fetchDynamicNodes(AwsEc2UnicastHostsProvider.java:117)
        at org.elasticsearch.discovery.ec2.AwsEc2UnicastHostsProvider$DiscoNodesCache.refresh(AwsEc2UnicastHostsProvider.java:232)
        at org.elasticsearch.discovery.ec2.AwsEc2UnicastHostsProvider$DiscoNodesCache.refresh(AwsEc2UnicastHostsProvider.java:217)
        at org.elasticsearch.common.util.SingleObjectCache.getOrRefresh(SingleObjectCache.java:54)
        at org.elasticsearch.discovery.ec2.AwsEc2UnicastHostsProvider.buildDynamicNodes(AwsEc2UnicastHostsProvider.java:103)
        at org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing.sendPings(UnicastZenPing.java:344)
        at org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing.ping(UnicastZenPing.java:249)
        at org.elasticsearch.discovery.zen.ping.ZenPingService.ping(ZenPingService.java:106)
        at org.elasticsearch.discovery.zen.ping.ZenPingService.pingAndWait(ZenPingService.java:84)
        at org.elasticsearch.discovery.zen.ZenDiscovery.findMaster(ZenDiscovery.java:845)
        at org.elasticsearch.discovery.zen.ZenDiscovery.innerJoinCluster(ZenDiscovery.java:376)
        at org.elasticsearch.discovery.zen.ZenDiscovery.access$4500(ZenDiscovery.java:89)
        at org.elasticsearch.discovery.zen.ZenDiscovery$JoinThreadControl$1.run(ZenDiscovery.java:1166)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:392)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by dadoonet**
Fix EC2 discovery setting

Closes #18652

Follow up for #18662

We add some tests to check that settings are correctly applied.
Tests revealed that some checks were missing.

Another PR will come after for S3 repositories but it's a bit more complex for repositories.

### CODE REVIEW

**Comment by dadoonet** [comment]
@rjernst Could you have a look at this one please?

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class AwsEc2ServiceImplTests extends ESTestCase {
+
+    public void testAWSCredentialsWithSystemProviders() {
+        AWSCredentials expectedCredentials = null;
+
+        try {
```
Comment (rjernst) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java]: Same concern about reproducibility as in the other PR.

**Comment by dadoonet** [comment]
@rjernst I pushed also another commit to this PR. Let me know.

**Comment by rjernst** [comment]
@dadoonet I don't think your change addresses my concern. That is, the try/catch in the *test* means we don't know what we are testing.

**Comment by dadoonet** [comment]
> That is, the try/catch in the test means we don't know what we are testing.

I see. Here is what I tried:

* Add in `repository-s3/build.gradle`:

```
test {
  // this is needed for insecure plugins, remove if possible!
  systemProperty 'tests.artifact', project.name
  systemProperty 'AWS_ACCESS_KEY_ID', 'DUMMY_ACCESS_KEY'
  systemProperty 'AWS_SECRET_KEY_ID', 'DUMMY_SECRET_KEY'
}
```

* Change the test with:

```java
    public void testAWSCredentialsWithSystemProviders() {
        AWSCredentialsProvider credentialsProvider = InternalAwsS3Service.buildCredentials(logger, "", "");

        AWSCredentials credentials = credentialsProvider.getCredentials();
        assertThat(credentials.getAWSAccessKeyId(), is("DUMMY_ACCESS_KEY"));
        assertThat(credentials.getAWSSecretKey(), is("DUMMY_SECRET_KEY"));
    }
```

But this fails. Is this the way I'm supposed to define system properties with Gradle?

It fails with:

```
java.lang.AssertionError: 
Expected: is "DUMMY_ACCESS_KEY"
     but: was "<HERE IS MY OWN AWS ACCESS KEY>"
````

**Comment by dadoonet** [comment]
@rjernst FYI I added a similar commit (which fails) to the other PR: https://github.com/elastic/elasticsearch/pull/18703/commits/5e3503576cf139c6442c7318f31bf44e64d694d8

**Comment by dadoonet** [comment]
@rjernst If you have some spare cycle could you help me on that error?

**Comment by rjernst** [comment]
@dadoonet did you mean for those to be env vars, not sysprops?

**Comment by dadoonet** [comment]
> did you mean for those to be env vars, not sysprops?

Hmmm. So indeed, when using sysprops I should define:

```
  systemProperty 'aws.accessKeyId', 'DUMMY_ACCESS_KEY'
  systemProperty 'aws.secretKey', 'DUMMY_SECRET_KEY'
```

But it still fails because the Credential chain is for now:

```java
            credentials = new AWSCredentialsProviderChain(
                new EnvironmentVariableCredentialsProvider(),
                new SystemPropertiesCredentialsProvider(),
                new InstanceProfileCredentialsProvider()
            );
```

As I have local env var for S3, it fails with:

```
java.lang.AssertionError: 
Expected: is "DUMMY_ACCESS_KEY"
     but: was "<HERE IS MY OWN AWS ACCESS KEY>"
```

Obviously, if I change the order in the chain with:

```java
            credentials = new AWSCredentialsProviderChain(
                new SystemPropertiesCredentialsProvider(),
                new EnvironmentVariableCredentialsProvider(),
                new InstanceProfileCredentialsProvider()
            );
```

It works from the CLI because Sysprops have now precedence on other settings.

For the record it does not work from the IDE as the IDE seems to not define the Sysprops in that case.
Not sure if we can omit this test if run from the IDE?

So? What do you think?

Should I change the order in the chain to make that test pass?
Or detect as I was doing previously if there is env vars that key/secret should correspond to the env var.
If env var not set, then check that key/secret is `DUMMY_ACCESS_KEY/DUMMY_SECRET_KEY`.

Thoughts?

**Comment by rjernst** [comment]
I would not do the previous "detection". That was my entire concern: the test would work differently depending on the environment it was in (eg you have aws keys in your env and I do not). I had thought we exposed the ability to add environment variables to the randomized runner (it supports it in ant), but apparently I forgot to add that to the gradle hook.  Moving the check to sysprop first is fine with me, but I would personally also make this more transparent to the user. Perhaps they should choose that they want to pull creds from sysprops, env, or instance profile? Rather than us having an arbitrary fallback order...it won't fix the testing issue here, but I think it would be a good change in general.

In any case, to test this, I think we need to:
1. Add env vars support to our gradle wrapper around randomized runner
2. Put either sys props or env vars in the test task
3. Add another task to run a one off test with the opposite set of creds

To test instance profile works, I think we will need to have a fixture (and not even sure that can work...), but the env and sysprops at least seem testable.

**Comment by dadoonet** [comment]
> Moving the check to sysprop first is fine with me, but I would personally also make this more transparent to the user. Perhaps they should choose that they want to pull creds from sysprops, env, or instance profile? 

In term of coding and testability, I agree that it makes more sense.
In term of simplicity for the user, I believe that using a chain of credential providers is really nice. 

Because users won't have to define it in the YML file explicitly. It seems to be a common practice on AWS and [recommended by AWS](http://docs.aws.amazon.com/java-sdk/latest/developer-guide/credentials.html). 

I think BTW that we should use `DefaultAWSCredentialsProviderChain` instead of providing the detail of the chain ourselves. If we do, you can ignore most of the next comments because testing this will still be hard as it depends on the plaform we run the test on.

I'd like to do this in another PR. The question of testing it will still remain as it will depend on 4 possibilities:

> AWS credentials provider chain that looks for credentials in this order:
> * Environment Variables - AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY (RECOMMENDED since they are recognized by all the AWS SDKs and CLI except for .NET), or AWS_ACCESS_KEY and AWS_SECRET_KEY (only recognized by Java SDK)
> * Java System Properties - aws.accessKeyId and aws.secretKey
> * Credential profiles file at the default location (~/.aws/credentials) shared by all AWS SDKs and the AWS CLI
> * Credentials delivered through the Amazon EC2 container service if AWS_CONTAINER_CREDENTIALS_RELATIVE_URI" environment variable is set and security manager has permission to access the variable,
> Instance profile credentials delivered through the Amazon EC2 metadata service

But here, I don't think that we want to test that `DefaultAWSCredentialsProviderChain` is working well. Well, it's provided by a library and we have to trust it. What we want to test here is basically that `if/else` condition:

```java
        if (key.isEmpty() && secret.isEmpty()) {
             // Scenario 1
        } else {
             // Scenario 2
        }
```

I don't want to block this PR because of this and I propose to mark this test as `@Ignore` or remove it entirely for now and discuss it in another PR where we change to the credential chain to `DefaultAWSCredentialsProviderChain`.

Do you agree with that plan?

**Comment by rjernst** [comment]
Sure,  fixing the test as a follow up here is fine.

**Comment by dadoonet** [comment]
Great! I created #19556 as a follow up for this.

**Comment by dadoonet** [comment]
@rjernst I updated this PR. Could you give a final review for it?
Do you think I should rebase on master and squash the commits?

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    protected void launchAWSCredentialsWithElasticsearchSettingsTest(Settings settings, String expectedKey, String expectedSecret) {
+        AWSCredentialsProvider credentialsProvider = AwsEc2ServiceImpl.buildCredentials(logger, settings);
+
+        try {
+            AWSCredentials credentials = credentialsProvider.getCredentials();
+            assertThat(credentials.getAWSAccessKeyId(), is(expectedKey));
+            assertThat(credentials.getAWSSecretKey(), is(expectedSecret));
+        } catch (AmazonClientException e) {
+            fail("We should not have raised an exception as we have credentials");
```
Comment (rjernst) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java]: You are throwing away the stack trace here. Just have this method throw Exception, and the tests that call it as well.

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    public void testInvalidRegion() {
+        Settings settings = Settings.builder()
+            .put(AwsEc2Service.REGION_SETTING.getKey(), "does-not-exist")
+            .build();
+        try {
+            AwsEc2ServiceImpl.findEndpoint(logger, settings);
+            fail("An invalid region should raise an exception");
+        } catch (IllegalArgumentException e) {
```
Comment (rjernst) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/cloud/aws/AwsEc2ServiceImplTests.java]: You can use expectThrows()

**Comment by rjernst** [comment]
LGTM, I left some minor comments.

**Pull request merged**